parquet/src/main/scala/magnolify/parquet/ParquetField.scala

```diff
@@ -29,44 +29,59 @@ import org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName
 import org.apache.parquet.schema.Type.Repetition
 import org.apache.parquet.schema.{LogicalTypeAnnotation, Type, Types}
 
-import scala.annotation.implicitNotFound
-import scala.collection.concurrent
+import scala.annotation.{implicitNotFound, nowarn}
 import scala.collection.compat._
+import scala.collection.concurrent
 
 sealed trait ParquetField[T] extends Serializable {
 
-  @transient private lazy val schemaCache: concurrent.Map[UUID, Type] =
+  @transient private lazy val schemaCache: concurrent.Map[Boolean, concurrent.Map[UUID, Type]] =
     concurrent.TrieMap.empty
 
-  protected def buildSchema(cm: CaseMapper): Type
-  def schema(cm: CaseMapper): Type =
-    schemaCache.getOrElseUpdate(cm.uuid, buildSchema(cm))
+  protected def buildSchema(cm: CaseMapper, groupArrayFields: Boolean): Type
+
+  def schema(
+    cm: CaseMapper,
+    groupArrayFields: Boolean
+  ): Type = {
+    if (!schemaCache.contains(groupArrayFields)) {
+      schemaCache.put(groupArrayFields, concurrent.TrieMap.empty)
+    }
+
+    schemaCache(groupArrayFields).getOrElseUpdate(
+      cm.uuid,
+      buildSchema(cm, groupArrayFields)
+    )
+  }
 
-  val hasAvroArray: Boolean = false
   def fieldDocs(cm: CaseMapper): Map[String, String]
   def typeDoc: Option[String]
 
-  protected val isGroup: Boolean = false
+  protected def isGroup(groupArrayFields: Boolean): Boolean = false
   protected def isEmpty(v: T): Boolean
   protected final def nonEmpty(v: T): Boolean = !isEmpty(v)
 
-  def write(c: RecordConsumer, v: T)(cm: CaseMapper): Unit
-  def newConverter: TypeConverter[T]
+  def write(c: RecordConsumer, v: T)(cm: CaseMapper, groupArrayFields: Boolean): Unit
```

Comment:
so this solution now works by passing a precomputed `groupArrayFields` flag around to all `write` method.

It works, but I'm thinking that we may want to generalize this from a boolean flag into a `Map[String, _]`-typed configuration field (parsed from the Hadoop `Configuration` object passed to `ParquetType`).

This would give us more flexibility if we need to make any more write options configurable in future. As an example, if a user has any `LocalDate*` fields in their `ParquetType` case class, by default parquet-avro will convert them into a `local-timestamp-{millis, micros}` for Avro 1.11, but `timestamp-{millis, micros}` on Avro 1.8. We might want to make this behavior configurable to preserve schema compatibility across Avro upgrades.

Reply:
wdyt @RustedBones ?

Reply:
We can use a custom trait containing all desired write setting

Reply:
updated to use a trait!